File: app/src/main/java/br/com/wcisang/alojinha/ui/fragment/BestSellersFragment.java
package br.com.wcisang.alojinha.ui.fragment;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import br.com.wcisang.alojinha.R;
import br.com.wcisang.alojinha.model.Product;
import br.com.wcisang.alojinha.ui.activity.ProductDetailActivity;
import br.com.wcisang.alojinha.ui.adapter.IntroCategoryAdapter;
import br.com.wcisang.alojinha.ui.adapter.ProductAdapter;
import br.com.wcisang.alojinha.viewmodel.BestSellersFragmentViewModel;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by WCisang on 07/06/2018.
 */
public class BestSellersFragment extends Fragment {


    private BestSellersFragmentViewModel viewModel;

    @BindView(R.id.recyclerview_product)
    RecyclerView recyclerView;

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.seller_fragment, container, false);
        ButterKnife.bind(this,v);
        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = ViewModelProviders.of(this).get(BestSellersFragmentViewModel.class);
        viewModel.getListMutableLiveData().observe(this, getBestSellersObserver());
        viewModel.init();
    }

    private Observer<List<Product>> getBestSellersObserver() {
        return this::setupList;
    }

    private void setupList(List<Product> products){
        progressBar.setVisibility(View.GONE);
        LinearLayoutManager manager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(manager);
        ProductAdapter adapter = new ProductAdapter(products, this::callDetail);
        recyclerView.setAdapter(adapter);
    }

    private void callDetail(Product product){
        Intent it = new Intent(getActivity(), ProductDetailActivity.class);
        it.putExtra(ProductDetailActivity.PRODUCT, product);
        startActivity(it);
    }
}

File: o-lodjinha/app/src/main/java/com/sumiya/olodjinha/UI/Fragments/BestSellersFragment.kt
package com.sumiya.olodjinha.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.ViewCompat
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.sumiya.olodjinha.R
import com.sumiya.olodjinha.contracts.ViewBestSellersFragmentContract
import com.sumiya.olodjinha.model.ProductDataModel
import com.sumiya.olodjinha.model.ProductModel
import com.sumiya.olodjinha.presenter.BestSellersPresenter
import com.sumiya.olodjinha.ui.adapter.BestSellersAdapter
import kotlinx.android.synthetic.main.fragment_best_sellers.*

class BestSellersFragment : Fragment(), ViewBestSellersFragmentContract {
    //Variables
    private var listener: BestSellersListener? = null

    //Lifecycle
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_best_sellers, container, false)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is BestSellersListener) {
            listener = context
        } else {
            throw RuntimeException(context.toString() + " must implement BestSellersListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        val bestSellersPresenter = BestSellersPresenter(this)
        bestSellersPresenter.getBestSellers()
    }

    //Contract
    override fun setSalesResponse(products: ProductDataModel) {
        bestSellersRecycler.adapter = BestSellersAdapter(
                products,
                { product: ProductModel -> productClicked(product) }
        )

        bestSellersRecycler.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        ViewCompat.setNestedScrollingEnabled(bestSellersRecycler, false)
        bestSellersRecycler.addItemDecoration(DividerItemDecoration(context!!, LinearLayoutManager.VERTICAL))
    }

    // Listeners
    private fun productClicked(product: ProductModel) {
        listener!!.openProductdetail(product)
    }

    //Fragment Listener
    interface BestSellersListener {
        fun openProductdetail(product: ProductModel)
    }
}
